File: DroidsAPI/src/main/java/com/github/rochedo098/droidsapi/api/script/classes/Robot.java
package com.github.rochedo098.droidsapi.api.script.classes;

import com.github.rochedo098.droidsapi.api.script.ScriptPath;
import com.github.rochedo098.droidsapi.api.script.classes.listeners.RobotListener;
import dev.latvian.mods.rhino.ScriptableObject;
import dev.latvian.mods.rhino.annotations.JSFunction;
import dev.latvian.mods.rhino.annotations.JSGetter;

import java.util.ArrayList;
import java.util.List;

public class Robot extends ScriptableObject {
    private final List<RobotListener> listeners = new ArrayList<>();

    public void addListener(RobotListener toAdd) {
        listeners.add(toAdd);
    }

    @Override
    public String getClassName() { return "Robot"; }

    @JSFunction
    public void register(String name, String fileLocation) {
        for (RobotListener rl : listeners) {
            rl.onRegister(name, ScriptPath.getLocation(fileLocation));
        }
    }

    @JSGetter
    public List<String> get(String identifier) {
        List<String> output = new ArrayList<>();
        for (RobotListener rl : listeners) {
            output.add(rl.get(identifier));
        }
        return output;
    }
}

File: src/main/kotlin/com/github/rochedo098/droids/api/script/classes/Robot.kt
package com.github.rochedo098.droids.api.script.classes

import com.github.rochedo098.droids.api.script.ScriptPath
import com.github.rochedo098.droids.api.script.classes.listeners.RobotListener
import dev.latvian.mods.rhino.ScriptableObject
import dev.latvian.mods.rhino.annotations.JSFunction
import dev.latvian.mods.rhino.annotations.JSGetter

class Robot : ScriptableObject() {
    private val listeners: MutableList<RobotListener> = ArrayList<RobotListener>()
    fun addListener(toAdd: RobotListener) { listeners.add(toAdd) }

    override fun getClassName(): String = "Robot"

    @JSFunction
    fun register(name: String?, fileLocation: String?) {
        for (rl in listeners) {
            rl.onRegister(name, fileLocation?.let { ScriptPath.getLocation(it) })
        }
    }

    @JSGetter
    operator fun get(identifier: String?): List<String> {
        val output: MutableList<String> = ArrayList()
        for (rl in listeners) {
            rl[identifier]?.let { output.add(it) }
        }
        return output
    }
}